### PR TITLE
Add basic envtest env setup and test coverage

### PR DESCRIPTION
**Pull request by gibizer**

### CODE REVIEW

**Comment by gibizer** [comment]
I need to tweak the makefile a bit to allow the unit test job to run the tests and pass in prow...

**Review by fultonj — COMMENTED** [review]
+2 I've used `make test` to run the env tests proposed here.

As per our learning session call this morning  we talked about the commits in this PR in detail. Thanks @gibizer 

I'll +2 so that a second reviewer can formally approve.

**Comment by gibizer** [comment]
> I need to tweak the makefile a bit to allow the unit test job to run the tests and pass in prow...

We need to land https://github.com/openshift/release/pull/39961 first to adapt the prow job for envtest.

**Comment by gibizer** [comment]
> > I need to tweak the makefile a bit to allow the unit test job to run the tests and pass in prow...
> 
> We need to land [openshift/release#39961](https://github.com/openshift/release/pull/39961) first to adapt the prow job for envtest.

And for that we need a small fix for the makefile first :/ https://github.com/openstack-k8s-operators/glance-operator/pull/240

File: Makefile [hunk 1]
```diff
@@ -131,9 +132,13 @@ golangci-lint:
...
 
+PROCS?=$(shell expr $(shell nproc --ignore 2) / 2)
+PROC_CMD = --procs ${PROCS}
+
 .PHONY: test
-test: manifests generate fmt vet envtest ## Run tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out
+test: manifests generate fmt vet envtest ginkgo ## Run tests.
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" \
+	 $(GINKGO) --trace --randomize-all ${PROC_CMD} $(GINKGO_ARGS) ./tests/functional/...
```
Comment (fmount) [inline Makefile:141]: +1 on switching to `ginkgo cli` as we've done in the other operators (FYI: same thing in progress in Manila [1])

[1] https://github.com/openstack-k8s-operators/manila-operator/pull/77

File: tests/functional/suite_test.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+		conn, err := tls.DialWithDialer(dialer, "tcp", addrPort, &tls.Config{InsecureSkipVerify: true})
+		if err != nil {
+			return err
+		}
+		conn.Close()
+		return nil
+	}).Should(Succeed())
+
+})
+
```
Comment (fmount) [inline tests/functional/suite_test.go:198]: One thing we could do here is to provide a `namespace` as part of the `BeforeEach` executed when the environment is initialized, but I understand you did it in the `glance_controller_test` for educational purposes. 
We can follow up on that part, it really depends on the scope of the tests.

File: tests/functional/suite_test.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
```
Comment (fmount) [inline tests/functional/suite_test.go:43]: we might want to add `networkv1` and the related resources to test that part too, but really a `TODO` list. The basic structure looks good to me.

**Comment by fmount** [comment]
@gibizer Thanks for this change.

The patch works as expected:

```
Will run 4 of 4 specs
Running in parallel across 5 processes
••••

Ran 4 of 4 Specs in 7.991 seconds
SUCCESS! -- 4 Passed | 0 Failed | 0 Pending | 0 Skipped


Ginkgo ran 1 suite in 10.317297719s
Test Suite Passed
```

Before merging I suspect we need to:

1. Rebase `go.mod`
2. Merge https://github.com/openshift/release/pull/39961 (which is ok to have it first as no tests are executed if not present)
3. Rekick the functional tests here once the previous one merges and double check everything is green (I think so given I tried it locally).
4. Squash (github should be able to do that for you?) and merge

File: tests/functional/suite_test.go (hunk 3 above)
Comment (gibizer) [reply tests/functional/suite_test.go:43]: I agree.

**Comment by gibizer** [comment]
I'm not sure what is happening in prow.

```
  [FAILED] in [BeforeSuite] - /go/src/github.com/openstack-k8s-operators/glance-operator/tests/functional/suite_test.go:114 @ 06/05/23 10:19:44.935
  << Timeline

  [FAILED] Unexpected error:
      <*fmt.wrapError | 0xc000140220>: 
      unable to install CRDs onto control plane: unable to create CRD instances: unable to create CRD "mariadbs.mariadb.openstack.org": Post "https://127.0.0.1:46171/apis/apiextensions.k8s.io/v1/customresourcedefinitions": http2: client connection lost
      {
          msg: "unable to install CRDs onto control plane: unable to create CRD instances: unable to create CRD \"mariadbs.mariadb.openstack.org\": Post \"https://127.0.0.1:46171/apis/apiextensions.k8s.io/v1/customresourcedefinitions\": http2: client connection lost",
          err: <*fmt.wrapError | 0xc000140200>{
              msg: "unable to create CRD instances: unable to create CRD \"mariadbs.mariadb.openstack.org\": Post \"https://127.0.0.1:46171/apis/apiextensions.k8s.io/v1/customresourcedefinitions\": http2: client connection lost",
              err: <*fmt.wrapError | 0xc0001401e0>{
                  msg: "unable to create CRD \"mariadbs.mariadb.openstack.org\": Post \"https://127.0.0.1:46171/apis/apiextensions.k8s.io/v1/customresourcedefinitions\": http2: client connection lost",
                  err: <*url.Error | 0xc000806000>{
                      Op: "Post",
                      URL: "https://127.0.0.1:46171/apis/apiextensions.k8s.io/v1/customresourcedefinitions",
                      Err: <*errors.errorString | 0xc0008c43c0>{
                          s: "http2: client connection lost",
                      },
                  },
              },
          },
      }
```

**Comment by fmount** [comment]
> I'm not sure what is happening in prow.
> 
> ```
>   [FAILED] in [BeforeSuite] - /go/src/github.com/openstack-k8s-operators/glance-operator/tests/functional/suite_test.go:114 @ 06/05/23 10:19:44.935
>   << Timeline
> 
>   [FAILED] Unexpected error:
>       <*fmt.wrapError | 0xc000140220>: 
>       unable to install CRDs onto control plane: unable to create CRD instances: unable to create CRD "mariadbs.mariadb.openstack.org": Post "https://127.0.0.1:46171/apis/apiextensions.k8s.io/v1/customresourcedefinitions": http2: client connection lost
>       {
>           msg: "unable to install CRDs onto control plane: unable to create CRD instances: unable to create CRD \"mariadbs.mariadb.openstack.org\": Post \"https://127.0.0.1:46171/apis/apiextensions.k8s.io/v1/customresourcedefinitions\": http2: client connection lost",
>           err: <*fmt.wrapError | 0xc000140200>{
>               msg: "unable to create CRD instances: unable to create CRD \"mariadbs.mariadb.openstack.org\": Post \"https://127.0.0.1:46171/apis/apiextensions.k8s.io/v1/customresourcedefinitions\": http2: client connection lost",
>               err: <*fmt.wrapError | 0xc0001401e0>{
>                   msg: "unable to create CRD \"mariadbs.mariadb.openstack.org\": Post \"https://127.0.0.1:46171/apis/apiextensions.k8s.io/v1/customresourcedefinitions\": http2: client connection lost",
>                   err: <*url.Error | 0xc000806000>{
>                       Op: "Post",
>                       URL: "https://127.0.0.1:46171/apis/apiextensions.k8s.io/v1/customresourcedefinitions",
>                       Err: <*errors.errorString | 0xc0008c43c0>{
>                           s: "http2: client connection lost",
>                       },
>                   },
>               },
>           },
>       }
> ```

This is really weird and actually unrelated to the patch. It seems that suite_test is failing during the setup, but not sure why.
I'll re-trigger  the job to see if it's a consistent failure at the same point.

**Comment by fmount** [comment]
@gibizer this time looks different [1]. I think it failed in the `BeforeEach` but I'm not sure if the real reason is some missing resources there. I'll take a better look at the errors during the day.

[1] https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_glance-operator/239/pull-ci-openstack-k8s-operators-glance-operator-main-functional/1665707514756337664/artifacts/test/build-log.txt

**Comment by gibizer** [comment]
It almost feels like that this job runs in a very different environment in prow than the rest of the functional jobs running envtest tests.

**Comment by gibizer** [comment]
Pushed https://github.com/openshift/release/pull/40018

**Comment by fmount** [comment]
> Pushed [openshift/release#40018](https://github.com/openshift/release/pull/40018)

Merging this one ^, let's see how it goes next.

**Comment by gibizer** [comment]
Now the tests looks good https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_glance-operator/239/pull-ci-openstack-k8s-operators-glance-operator-main-functional/1665814367788077056/artifacts/test/build-log.txt

**Comment by fmount** [comment]
> Now the tests looks good https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_glance-operator/239/pull-ci-openstack-k8s-operators-glance-operator-main-functional/1665814367788077056/artifacts/test/build-log.txt

Yes, so at least we found a constraint about the minimum requirements to run the job in prow. At this point I think we can merge it and follow up (thanks for starting this).

**Review by fmount — APPROVED** [review]

**Pull request merged**